spacy_experimental/coref/coref_component.py

```diff
@@ -0,0 +1,395 @@
+        name (str): The component instance name, used to add entries to the
+            losses during training.
+        span_mentions (str): Key in doc.spans where the candidate coref mentions
+            are stored in.
+        span_cluster_prefix (str): Prefix for the key in doc.spans to store the
+            coref clusters in.
+        scorer (Optional[Callable]): The scoring method. Defaults to
+            Scorer.score_coref_clusters.
+
+        DOCS: https://spacy.io/api/coref#init
+        """
+        self.vocab = vocab
+        self.model = model
+        self.name = name
+        self.span_mentions = span_mentions
+        self.span_cluster_prefix = span_cluster_prefix
+        self._rehearsal_model = None
+
+        self.cfg: Dict[str, Any] = {"span_cluster_prefix": span_cluster_prefix}
+        self.scorer = scorer
+
+    def predict(self, docs: Iterable[Doc]) -> List[MentionClusters]:
+        """Apply the pipeline's model to a batch of docs, without modifying them.
+        Return the list of predicted clusters.
+
+        docs (Iterable[Doc]): The documents to predict.
+        RETURNS (List[MentionClusters]): The model's prediction for each document.
+
+        DOCS: https://spacy.io/api/coref#predict
+        """
+        out = []
+        for doc in docs:
+            scores, idxs = self.model.predict([doc])
+            # idxs is a list of mentions (start / end idxs)
+            # each item in scores includes scores and a mapping from scores to mentions
+            ant_idxs = idxs
+
+            # TODO batching
+            xp = self.model.ops.xp
+
+            starts = xp.arange(0, len(doc))
+            ends = xp.arange(0, len(doc)) + 1
+
+            predicted = get_predicted_clusters(xp, starts, ends, ant_idxs, scores)
+            out.append(predicted)
+
+        return out
+
+    def set_annotations(self, docs: Iterable[Doc], clusters_by_doc) -> None:
+        """Modify a batch of Doc objects, using pre-computed scores.
+
+        docs (Iterable[Doc]): The documents to modify.
+        clusters: The span clusters, produced by CoreferenceResolver.predict.
+
+        DOCS: https://spacy.io/api/coref#set_annotations
+        """
+        docs = list(docs)
+        if len(docs) != len(clusters_by_doc):
+            raise ValueError(
+                "Found coref clusters incompatible with the "
+                "documents provided to the 'coref' component. "
+                "This is likely a bug in spaCy."
+            )
+        for doc, clusters in zip(docs, clusters_by_doc):
+            for ii, cluster in enumerate(clusters):
+                key = self.span_cluster_prefix + "_" + str(ii)
+                if key in doc.spans:
+                    raise ValueError(
+                        "Found coref clusters incompatible with the "
+                        "documents provided to the 'coref' component. "
+                        "This is likely a bug in spaCy."
```

Comment:
That's a good point - I think it came up before, I'll see if I can find the old discussion.

Reply:
(I will also fix the error message)

Reply:
For the revised error message, how about this?

 > Existing spans found, not overwriting. If you run more than one instance of coref you should use different target span prefixes.

Reply:
I found the old discussion, I mentioned maybe adding a warning for leftover spans but didn't follow up on it.

I do think it would be good to have a warning, but the check should be minimal. If we only check for spans we're setting the cost should be small enough, but if we iterate over every span of every Doc in every batch I think that could get costly. Because the span keys are generated from a template, checking if they match the template requires either:

1. iterating over numbers to do string comparisons, or 
2. splitting the key and then checking the parts

Both of which don't seem great. We do can do things like filter on prefix but then we still have to do a partial string match on every span key.

Is there a faster way to check for strays like that?

Reply:
Can't we just iterate over the keys in `doc.spans` and check whether any of them start with `span_cluster_prefix + "_"`?

Reply:
We actually can't do that, because for example if the prefix is `xxx` then `xxx_1` could be a coref piece of data, but `xxx_yyy_1` would be valid and unrelated.

I may be worrying about the cost of this too much, since I guess it's not typical to have many SpanGroups, but it is something that came to mind.

Reply:
But we could check all the keys in `doc.spans` to follow the pattern `span_cluster_prefix` + `_` + some integer? Doing that splitting & trying to parse as integer shouldn't take much time, if we only do it once for the existing keys?

Reply:
I also think you wouldn't typically have thousands of keys, right?

Reply:
I think you're right and I was too worried about the performance. I went ahead and implemented this [here](https://github.com/explosion/spacy-experimental/pull/17/commits/1e7fdf746998f3524a45c4d3289c6b50f5d24b45). I still have doubts about whether this is the right way to do this, because it seems like it could be very noisy, or we could check for these annotations before doing inference, rather than when adding annotations. I'll leave this thread open and think about it some more.

Reply:
Thinking about this more, I think it's better to leave the error (which checks for actual overwrite) and remove the warning (which checks for potential overwrite). 

If we have both the error and the warning, then cases where the error will be triggered will also trigger the warning, which is unhelpful and confusing. On the other hand, I think a warning isn't strong enough for actually overwriting annotations in an unusual way. (I could do something like check the warning and only raise it if the error is not fired, but that also seemed overly complicated.)

One thing we could do is to change the current warning to be an error instead and drop the error. That seems heavy-handed though. 

Part of why I think this is that it's pretty easy to imagine running the same coref component, or similar components, twice in a row, but I have trouble imagining a situation where high-numbered spans that won't get overwritten exist, but low-numbered spans don't. That would require not only running multiple components with the same prefix, but also partly removing span groups.

Reply:
Ok. Let's error on actual overwrite (like I think you had it originally...)

Reply:
I think this has been addressed, so resolving.